File: src/daemon.c
/* SPDX-License-Identifier: MIT */

#include "hyprstream.h"

#include <errno.h>
#include <poll.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <unistd.h>

static struct hs_state g_state;

static void reconcile(struct hs_state *st);
static int  streaming_enable(struct hs_state *st);
static int  streaming_disable(struct hs_state *st);

static void handle_signal(int sig)
{
    (void)sig;
    g_state.running = false;
}

static int streaming_enable(struct hs_state *st)
{
    if (st->mode == HS_MODE_ENABLED) {
        hs_info("already enabled");
        return 0;
    }

    if (st->physical_monitor[0] == '\0') {
        if (hs_detect_physical_monitor(st->physical_monitor,
                                       sizeof(st->physical_monitor)) < 0)
            return -1;
    }

    char *before_workspaces = hs_hyprctl_json("workspaces");

    if (hs_create_headless(st->headless_name, sizeof(st->headless_name)) < 0) {
        free(before_workspaces);
        return -1;
    }

    hs_disable_mirror(st->headless_name, st->cfg.virtual_resolution);
    if (hs_bind_workspace_to_monitor(st->cfg.streaming_workspace, st->headless_name) < 0)
        goto fail;
    if (hs_move_workspace_to_monitor(st->cfg.streaming_workspace, st->headless_name) < 0)
        goto fail;

    /*
     * Hyprland may assign an existing workspace to a newly-created monitor.
     * If that steals a workspace from a physical monitor, restore it back.
     */
    hs_restore_headless_stolen_workspaces(before_workspaces,
                                          st->headless_name,
                                          st->cfg.streaming_workspace);
    free(before_workspaces);

    st->mode = HS_MODE_ENABLED;
    st->mirroring_active = false;
    st->on_streaming_workspace = false;

    reconcile(st);

    if (st->cfg.on_enable[0])
        hs_exec_async(st->cfg.on_enable);

    hs_info("streaming mode enabled (headless=%s, physical=%s)",
            st->headless_name, st->physical_monitor);
    return 0;

fail:
    hs_restore_headless_stolen_workspaces(before_workspaces,
                                          st->headless_name,
                                          st->cfg.streaming_workspace);
    free(before_workspaces);
    if (st->headless_name[0] != '\0') {
        hs_remove_headless(st->headless_name);
        st->headless_name[0] = '\0';
    }
    return -1;
}

static int streaming_disable(struct hs_state *st)
{
    if (st->mode == HS_MODE_DISABLED) {
        hs_info("already disabled");
        return 0;
    }

    if (st->mirroring_active) {
        hs_disable_mirror(st->headless_name, st->cfg.virtual_resolution);
        st->mirroring_active = false;
    }

    hs_move_workspace_to_monitor(st->cfg.streaming_workspace, st->physical_monitor);

    if (st->headless_name[0] != '\0') {
        hs_remove_headless(st->headless_name);
        st->headless_name[0] = '\0';
    }

    st->mode = HS_MODE_DISABLED;
    st->on_streaming_workspace = false;

    if (st->cfg.on_disable[0])
        hs_exec_async(st->cfg.on_disable);

    hs_info("streaming mode disabled");
    return 0;
}

static void reconcile(struct hs_state *st)
{
    if (st->mode != HS_MODE_ENABLED)
        return;

    char ws[HS_MAX_NAME];
    if (hs_get_active_workspace(ws, sizeof(ws)) < 0)
        return;

    bool on_stream = (strcmp(ws, st->cfg.streaming_workspace) == 0);

    if (on_stream && !st->mirroring_active) {
        hs_info("entering streaming workspace -> enable mirror");
        hs_enable_mirror(st->headless_name, st->physical_monitor);
        st->mirroring_active = true;
        st->on_streaming_workspace = true;

        if (st->cfg.on_streaming_enter[0])
            hs_exec_async(st->cfg.on_streaming_enter);

    } else if (!on_stream && st->mirroring_active) {
        hs_info("leaving streaming workspace -> disable mirror");
        hs_disable_mirror(st->headless_name, st->cfg.virtual_resolution);
        st->mirroring_active = false;
        st->on_streaming_workspace = false;

        if (st->cfg.on_streaming_leave[0])
            hs_exec_async(st->cfg.on_streaming_leave);
    }

    hs_strlcpy(st->active_workspace, ws, sizeof(st->active_workspace));
}

static void on_hyprland_event(const char *event, const char *data, void *ctx)
{
    struct hs_state *st = ctx;
    (void)data;

    if (strcmp(event, "workspace") == 0 ||
        strcmp(event, "focusedmon") == 0 ||
        strcmp(event, "activewindow") == 0 ||
        strcmp(event, "movewindow") == 0) {
        reconcile(st);
    } else if (strcmp(event, "monitorremoved") == 0) {
        if (st->mode == HS_MODE_ENABLED &&
            st->headless_name[0] != '\0' &&
            strcmp(data, st->headless_name) == 0) {
            hs_warn("headless output was removed externally");
            st->mode = HS_MODE_DISABLED;
            st->headless_name[0] = '\0';
            st->mirroring_active = false;
            st->on_streaming_workspace = false;
        }
    }
}

static void handle_ctl_client(int client_fd, struct hs_state *st)
{
    char buf[256];
    ssize_t n = read(client_fd, buf, sizeof(buf) - 1);
    if (n <= 0) {
        close(client_fd);
        return;
    }
    buf[n] = '\0';
    hs_trim(buf);

    char response[1024];

    if (strcmp(buf, HS_CTL_ENABLE) == 0) {
        int rc = streaming_enable(st);
        snprintf(response, sizeof(response),
                 rc == 0 ? "enabled" : "error: enable failed");

    } else if (strcmp(buf, HS_CTL_DISABLE) == 0) {
        int rc = streaming_disable(st);
        snprintf(response, sizeof(response),
                 rc == 0 ? "disabled" : "error: disable failed");

    } else if (strcmp(buf, HS_CTL_TOGGLE) == 0) {
        int rc;
        if (st->mode == HS_MODE_ENABLED)
            rc = streaming_disable(st);
        else
            rc = streaming_enable(st);
        snprintf(response, sizeof(response),
                 rc == 0 ? (st->mode == HS_MODE_ENABLED ? "enabled" : "disabled")
                         : "error: toggle failed");

    } else if (strcmp(buf, HS_CTL_STATUS) == 0) {
        snprintf(response, sizeof(response),
                 "mode=%s headless=%s physical=%s workspace=%s mirroring=%s",
                 st->mode == HS_MODE_ENABLED ? "enabled" : "disabled",
                 st->headless_name[0] ? st->headless_name : "none",
                 st->physical_monitor[0] ? st->physical_monitor : "unknown",
                 st->active_workspace,
                 st->mirroring_active ? "on" : "off");

    } else if (strcmp(buf, HS_CTL_QUIT) == 0) {
        snprintf(response, sizeof(response), "shutting down");
        st->running = false;

    } else {
        snprintf(response, sizeof(response), "error: unknown command: %s", buf);
    }

    ssize_t wr = write(client_fd, response, strlen(response));
    (void)wr;
    close(client_fd);
}

int hs_daemon_run(struct hs_config *cfg)
{
    memset(&g_state, 0, sizeof(g_state));
    g_state.cfg = *cfg;
    g_state.running = true;
    g_state.mode = HS_MODE_DISABLED;

    signal(SIGINT, handle_signal);
    signal(SIGTERM, handle_signal);
    signal(SIGPIPE, SIG_IGN);

    if (cfg->physical_monitor[0]) {
        hs_strlcpy(g_state.physical_monitor, cfg->physical_monitor,
                    sizeof(g_state.physical_monitor));
    } else {
        if (hs_detect_physical_monitor(g_state.physical_monitor,
                                       sizeof(g_state.physical_monitor)) < 0) {
            hs_error("could not detect physical monitor");
            return 1;
        }
    }

    int ctl_fd = hs_ctl_create();
    if (ctl_fd < 0)
        return 1;
    g_state.ctl_fd = ctl_fd;

    if (cfg->auto_enable) {
        if (streaming_enable(&g_state) < 0)
            hs_warn("auto-enable failed, continuing in disabled mode");
    }

    int reconnect_attempts = 0;

    while (g_state.running) {
        int ipc_fd = hs_ipc_connect();
        if (ipc_fd < 0) {
            reconnect_attempts++;
            if (reconnect_attempts > HS_MAX_RECONNECT) {
                hs_error("max reconnect attempts reached, exiting");
                break;
            }
            hs_warn("IPC connect failed, retrying in %ds (%d/%d)",
                    HS_RECONNECT_DELAY, reconnect_attempts, HS_MAX_RECONNECT);
            sleep(HS_RECONNECT_DELAY);
            continue;
        }
        reconnect_attempts = 0;
        g_state.ipc_fd = ipc_fd;

        struct pollfd fds[2];
        fds[0].fd = ipc_fd;
        fds[0].events = POLLIN;
        fds[1].fd = ctl_fd;
        fds[1].events = POLLIN;

        while (g_state.running) {
            int ret = poll(fds, 2, 1000);
            if (ret < 0) {
                if (errno == EINTR)
                    continue;
                hs_error("poll: %s", strerror(errno));
                break;
            }

            if (fds[0].revents & POLLIN) {
                char buf[HS_IPC_BUF];
                ssize_t n = read(ipc_fd, buf, sizeof(buf) - 1);
                if (n <= 0) {
                    hs_warn("IPC connection lost, reconnecting...");
                    break;
                }
                buf[n] = '\0';

                char *line = buf;
                char *nl;
                while ((nl = strchr(line, '\n')) != NULL) {
                    *nl = '\0';
                    char *sep = strstr(line, ">>");
                    if (sep) {
                        *sep = '\0';
                        on_hyprland_event(line, sep + 2, &g_state);
                    }
                    line = nl + 1;
                }
            }

            if (fds[0].revents & (POLLHUP | POLLERR)) {
                hs_warn("IPC socket error, reconnecting...");
                break;
            }

            if (fds[1].revents & POLLIN) {
                int client = accept(ctl_fd, NULL, NULL);
                if (client >= 0)
                    handle_ctl_client(client, &g_state);
            }
        }

        close(ipc_fd);
        g_state.ipc_fd = -1;
    }

    if (g_state.mode == HS_MODE_ENABLED)
        streaming_disable(&g_state);

    close(ctl_fd);
    unlink(hs_ctl_socket_path());

    hs_info("daemon stopped");
    return 0;
}

File: src/daemon.rs
use crate::config::Config;
use crate::control::Command as CtlCommand;
use crate::hypr;
use crate::paths;
use crate::{hs_error, hs_info, hs_warn};
use anyhow::{Context, Result};
use libc::{poll, pollfd, POLLERR, POLLHUP, POLLIN};
use std::collections::HashMap;
use std::fs;
use std::io::{Read, Write};
use std::os::unix::fs::PermissionsExt;
use std::os::unix::io::AsRawFd;
use std::os::unix::net::{UnixListener, UnixStream};
use std::process::Command as ProcessCommand;
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc,
};
use std::time::Duration;

const RECONNECT_DELAY_SECS: u64 = 2;
const MAX_RECONNECT: usize = 30;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Mode {
    Disabled,
    Enabled,
}

struct State {
    mode: Mode,
    headless: Option<String>,
    physical: String,
    active_workspace: String,
    mirroring_active: bool,
    cfg: Config,
}

fn exec_async(cmd: &str) {
    if cmd.trim().is_empty() {
        return;
    }
    let _ = ProcessCommand::new("sh").arg("-c").arg(cmd).spawn();
}

#[cfg(test)]
mod tests {
    use super::{reconcile, Mode, State};
    use crate::config::Config;
    use crate::hypr::Hypr;
    use anyhow::{anyhow, Result};
    use serde::de::DeserializeOwned;
    use serde_json::json;
    use std::cell::RefCell;

    struct FakeHypr {
        active_ws: RefCell<String>,
        raw_calls: RefCell<Vec<String>>,
    }

    impl FakeHypr {
        fn new(active_ws: &str) -> Self {
            Self {
                active_ws: RefCell::new(active_ws.to_string()),
                raw_calls: RefCell::new(Vec::new()),
            }
        }

        fn set_active(&self, ws: &str) {
            *self.active_ws.borrow_mut() = ws.to_string();
        }
    }

    impl Hypr for FakeHypr {
        fn request_raw(&self, args: &str) -> Result<String> {
            self.raw_calls.borrow_mut().push(args.to_string());
            Ok("ok".to_string())
        }

        fn request_json<T: DeserializeOwned>(&self, args: &str) -> Result<T> {
            if args == "activeworkspace" {
                let name = self.active_ws.borrow().clone();
                let v = json!({"name": name});
                return Ok(serde_json::from_value::<T>(v)?);
            }
            Err(anyhow!("unsupported json query in test: {args}"))
        }
    }

    #[test]
    fn reconcile_enables_and_disables_mirror_on_workspace_change() {
        let cfg = Config::default();
        let fake = FakeHypr::new(&cfg.streaming_workspace);
        let mut st = State {
            mode: Mode::Enabled,
            headless: Some("HEADLESS-1".to_string()),
            physical: "eDP-1".to_string(),
            active_workspace: String::new(),
            mirroring_active: false,
            cfg,
        };

        reconcile(&mut st, &fake).unwrap();
        assert!(st.mirroring_active);
        assert!(fake
            .raw_calls
            .borrow()
            .iter()
            .any(|c| c.contains("mirror,eDP-1")));

        fake.set_active("1");
        reconcile(&mut st, &fake).unwrap();
        assert!(!st.mirroring_active);
        assert!(fake
            .raw_calls
            .borrow()
            .iter()
            .any(|c| c.contains("-9999x0")));
    }
}

fn reconcile(st: &mut State, ipc: &impl hypr::Hypr) -> Result<()> {
    if st.mode != Mode::Enabled {
        return Ok(());
    }
    let headless = match st.headless.as_deref() {
        Some(h) => h,
        None => return Ok(()),
    };

    let ws = hypr::active_workspace(ipc)?;
    let on_stream = ws == st.cfg.streaming_workspace;

    if on_stream && !st.mirroring_active {
        hs_info!("entering streaming workspace -> enable mirror");
        hypr::enable_mirror(ipc, headless, &st.physical)?;
        st.mirroring_active = true;
        if !st.cfg.on_streaming_enter.is_empty() {
            exec_async(&st.cfg.on_streaming_enter);
        }
    } else if !on_stream && st.mirroring_active {
        hs_info!("leaving streaming workspace -> disable mirror");
        hypr::disable_mirror(ipc, headless, &st.cfg.virtual_resolution)?;
        st.mirroring_active = false;
        if !st.cfg.on_streaming_leave.is_empty() {
            exec_async(&st.cfg.on_streaming_leave);
        }
    }

    st.active_workspace = ws;
    Ok(())
}

fn streaming_enable(st: &mut State, ipc: &impl hypr::Hypr) -> Result<()> {
    if st.mode == Mode::Enabled {
        hs_info!("already enabled");
        return Ok(());
    }

    if st.physical.is_empty() {
        st.physical = if !st.cfg.physical_monitor.is_empty() {
            st.cfg.physical_monitor.clone()
        } else {
            hypr::detect_physical_monitor(ipc)?
        };
    }

    let before: HashMap<String, String> = hypr::snapshot_workspaces(ipc)?;
    let headless = hypr::create_headless(ipc)?;

    // park headless off-screen
    hypr::disable_mirror(ipc, &headless, &st.cfg.virtual_resolution)?;

    // bind/move streaming workspace to headless
    let res = (|| -> Result<()> {
        hypr::bind_workspace_to_monitor(ipc, &st.cfg.streaming_workspace, &headless)?;
        hypr::move_workspace_to_monitor(ipc, &st.cfg.streaming_workspace, &headless)?;
        Ok(())
    })();

    if let Err(e) = res {
        let _ = hypr::restore_headless_stolen_workspaces(
            ipc,
            &before,
            &headless,
            &st.cfg.streaming_workspace,
        );
        let _ = hypr::remove_headless(ipc, &headless);
        return Err(e);
    }

    let _ = hypr::restore_headless_stolen_workspaces(
        ipc,
        &before,
        &headless,
        &st.cfg.streaming_workspace,
    );

    st.mode = Mode::Enabled;
    st.headless = Some(headless);
    st.mirroring_active = false;
    st.active_workspace.clear();

    reconcile(st, ipc)?;

    if !st.cfg.on_enable.is_empty() {
        exec_async(&st.cfg.on_enable);
    }

    hs_info!(
        "streaming mode enabled (headless={:?}, physical={})",
        st.headless,
        st.physical
    );
    Ok(())
}

fn streaming_disable(st: &mut State, ipc: &impl hypr::Hypr) -> Result<()> {
    if st.mode == Mode::Disabled {
        hs_info!("already disabled");
        return Ok(());
    }

    if let Some(headless) = st.headless.as_deref() {
        if st.mirroring_active {
            hypr::disable_mirror(ipc, headless, &st.cfg.virtual_resolution)?;
            st.mirroring_active = false;
        }

        let _ = hypr::move_workspace_to_monitor(ipc, &st.cfg.streaming_workspace, &st.physical);
        let _ = hypr::remove_headless(ipc, headless);
    }

    st.headless = None;
    st.mode = Mode::Disabled;

    if !st.cfg.on_disable.is_empty() {
        exec_async(&st.cfg.on_disable);
    }

    hs_info!("streaming mode disabled");
    Ok(())
}

fn handle_ctl_with_running(
    client: UnixStream,
    st: &mut State,
    ipc: &impl hypr::Hypr,
    running: &Arc<AtomicBool>,
) -> Result<()> {
    // Peek command first so we can flip running for quit.
    let mut buf = [0u8; 256];
    let mut client = client;
    let n = client.read(&mut buf).context("read control command")?;
    if n == 0 {
        return Ok(());
    }
    let cmd_raw = String::from_utf8_lossy(&buf[..n]).trim().to_string();

    let response = match CtlCommand::parse(&cmd_raw) {
        Ok(CtlCommand::Enable) => {
            if streaming_enable(st, ipc).is_ok() {
                "enabled".to_string()
            } else {
                "error: enable failed".to_string()
            }
        }
        Ok(CtlCommand::Disable) => {
            if streaming_disable(st, ipc).is_ok() {
                "disabled".to_string()
            } else {
                "error: disable failed".to_string()
            }
        }
        Ok(CtlCommand::Toggle) => {
            let rc = if st.mode == Mode::Enabled {
                streaming_disable(st, ipc)
            } else {
                streaming_enable(st, ipc)
            };
            if rc.is_ok() {
                if st.mode == Mode::Enabled {
                    "enabled".to_string()
                } else {
                    "disabled".to_string()
                }
            } else {
                "error: toggle failed".to_string()
            }
        }
        Ok(CtlCommand::Status) => format!(
            "mode={} headless={} physical={} workspace={} mirroring={}",
            if st.mode == Mode::Enabled {
                "enabled"
            } else {
                "disabled"
            },
            st.headless.as_deref().unwrap_or("none"),
            if st.physical.is_empty() {
                "unknown"
            } else {
                &st.physical
            },
            st.active_workspace,
            if st.mirroring_active { "on" } else { "off" }
        ),
        Ok(CtlCommand::Quit) => "shutting down".to_string(),
        Err(_e) => format!("error: unknown command: {cmd_raw}"),
    };

    if cmd_raw.trim() == CtlCommand::Quit.as_str() {
        running.store(false, Ordering::Relaxed);
    }
    let _ = client.write_all(response.as_bytes());
    Ok(())
}

fn create_ctl_listener() -> Result<UnixListener> {
    let path = paths::ctl_socket_path();
    let _ = fs::remove_file(&path);
    let l = UnixListener::bind(&path)
        .with_context(|| format!("bind control socket: {}", path.display()))?;
    fs::set_permissions(&path, fs::Permissions::from_mode(0o700))?;
    hs_info!("control socket: {}", path.display());
    Ok(l)
}

fn connect_events() -> Result<UnixStream> {
    let p = paths::hypr_event_socket_path()?;
    let s = UnixStream::connect(&p)
        .with_context(|| format!("connect hyprland IPC: {}", p.display()))?;
    hs_info!("connected to hyprland IPC: {}", p.display());
    Ok(s)
}

fn handle_event_line(st: &mut State, ipc: &impl hypr::Hypr, line: &str) {
    let Some((event, data)) = line.split_once(">>") else {
        return;
    };

    match event {
        "workspace" | "focusedmon" | "activewindow" | "movewindow" => {
            if let Err(e) = reconcile(st, ipc) {
                hs_warn!("reconcile failed: {e:#}");
            }
        }
        "monitorremoved" => {
            if st.mode == Mode::Enabled {
                if let Some(headless) = st.headless.as_deref() {
                    if data == headless {
                        hs_warn!("headless output was removed externally");
                        st.mode = Mode::Disabled;
                        st.headless = None;
                        st.mirroring_active = false;
                    }
                }
            }
        }
        _ => {}
    }
}

pub fn run(cfg: Config) -> Result<()> {
    let ipc = hypr::HyprIpc::new()?;

    let running = Arc::new(AtomicBool::new(true));
    // Best-effort signal handling; if it fails we still run.
    let _ = signal_hook::flag::register(libc::SIGINT, Arc::clone(&running));
    let _ = signal_hook::flag::register(libc::SIGTERM, Arc::clone(&running));

    let ctl = create_ctl_listener()?;
    ctl.set_nonblocking(true)?;

    let mut st = State {
        mode: Mode::Disabled,
        headless: None,
        physical: String::new(),
        active_workspace: String::new(),
        mirroring_active: false,
        cfg,
    };

    if st.cfg.auto_enable {
        if let Err(e) = streaming_enable(&mut st, &ipc) {
            hs_warn!("auto-enable failed, continuing in disabled mode: {e:#}");
        }
    }

    let mut reconnect_attempts = 0usize;
    let mut buf: Vec<u8> = Vec::with_capacity(64 * 1024);

    while running.load(Ordering::Relaxed) {
        let mut events = match connect_events() {
            Ok(s) => s,
            Err(e) => {
                reconnect_attempts += 1;
                if reconnect_attempts > MAX_RECONNECT {
                    hs_error!("max reconnect attempts reached, exiting");
                    break;
                }
                hs_warn!(
                    "IPC connect failed, retrying in {}s ({}/{}) ({e:#})",
                    RECONNECT_DELAY_SECS,
                    reconnect_attempts,
                    MAX_RECONNECT
                );
                std::thread::sleep(Duration::from_secs(RECONNECT_DELAY_SECS));
                continue;
            }
        };
        reconnect_attempts = 0;
        events.set_nonblocking(true)?;

        let mut fds = [
            pollfd {
                fd: events.as_raw_fd(),
                events: POLLIN,
                revents: 0,
            },
            pollfd {
                fd: ctl.as_raw_fd(),
                events: POLLIN,
                revents: 0,
            },
        ];

        'inner: loop {
            if !running.load(Ordering::Relaxed) {
                break 'inner;
            }
            let rc = unsafe { poll(fds.as_mut_ptr(), fds.len() as libc::nfds_t, 1000) };
            if rc < 0 {
                let e = std::io::Error::last_os_error();
                if e.kind() == std::io::ErrorKind::Interrupted {
                    continue;
                }
                return Err(e).context("poll");
            }

            if (fds[0].revents & (POLLHUP | POLLERR)) != 0 {
                hs_warn!("IPC socket error, reconnecting...");
                break 'inner;
            }

            if (fds[0].revents & POLLIN) != 0 {
                let mut tmp = [0u8; 65536];
                match events.read(&mut tmp) {
                    Ok(0) => {
                        hs_warn!("IPC connection lost, reconnecting...");
                        break 'inner;
                    }
                    Ok(n) => {
                        buf.extend_from_slice(&tmp[..n]);
                        while let Some(pos) = buf.iter().position(|b| *b == b'\n') {
                            let line = buf.drain(..=pos).collect::<Vec<u8>>();
                            let line = String::from_utf8_lossy(&line);
                            let line = line.trim_end_matches(['\n', '\r']);
                            if !line.is_empty() {
                                handle_event_line(&mut st, &ipc, line);
                            }
                        }
                    }
                    Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {}
                    Err(e) => {
                        hs_warn!("IPC read error, reconnecting: {}", e);
                        break 'inner;
                    }
                }
            }

            if (fds[1].revents & POLLIN) != 0 {
                loop {
                    match ctl.accept() {
                        Ok((client, _addr)) => {
                            if let Err(e) = handle_ctl_with_running(client, &mut st, &ipc, &running)
                            {
                                hs_warn!("control handler failed: {e:#}");
                            }
                        }
                        Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => break,
                        Err(e) => return Err(e).context("accept control")?,
                    }
                }
            }
        }
    }

    if st.mode == Mode::Enabled {
        let _ = streaming_disable(&mut st, &ipc);
    }
    let _ = fs::remove_file(paths::ctl_socket_path());
    hs_info!("daemon stopped");
    Ok(())
}
